Revise and complete chapter on SPARK (#399)
AdaCore/learn: content/courses/Ada_For_The_C_Embedded_Developer/chapters/05_SPARK.rst

.. include:: ../../global.txt

Understanding Exceptions and Dynamic Checks
-------------------------------------------

In Ada, several common programming errors that are not already detected 
at compile-time are detected instead at run-time, triggering 
"exceptions" that interrupt the normal flow of execution. For example, 
an exception is raised by an attempt to access an array component via an 
index that is out of bounds, thereby precluding exploits based on buffer 

> overflow. Several other cases also raise language-defined exceptions, 

Reviewer: Editorial suggestion: you could split this sentence.

such as scalar range constraint violations, and users may declare and 
raise their own application-specific exceptions. 

We will loosely define a "frame" as a language construct that has a call
stack entry when called, for example a procedure or function body. There
are a few other constructs that are also pertinent but this definition
will suffice for now.

Frames can have "exception handlers" that specify the response to be
executed when an exception is "raised." These handlers can be specified to

> "handle" one or more specific exceptions, thus allowing the response to

Reviewer: Editorial suggestion: although this sentence is correct, it might be too formal / abstract for the more casual reader.

be tailored to specific exceptions or generalized to more than one.

Structurally, the exception handlers are grouped together and textually

> separated from the rest of the code in a frame. As a result, there is a

Reviewer: Editorial suggestion: maybe this whole paragraph (and maybe the previous one, too) should come _after_ the source-code example. In this case, you'd be able to indicate the distinct parts in the source-code. For example:

```suggestion
separated from the rest of the code in a frame, as you can see at line XX with the :ada:`exception` keyword.. As a result, there is a
```

sequence of statements representing the normal flow of execution,
distinct from another section representing the error handling. The
reserved word "exception" separates these two sections. This separation
helps simplify the overall flow, increasing understandability. In
particular, status result codes are not required so there is no mixing
of error checking and normal progress. If no exception is raised the
exception handler section is automatically skipped during execution.

For example, in the following procedure line 4 separates the two
sections. Lines 5 through 10 contain the error handling part,

> consisting in this case of three exception handlers:

Reviewer: Just as a suggestion:
```suggestion
consisting in this case of three exception handlers. Each exception handler starts with the :ada:`when` keyword.
```

:code-config:`accumulate_code=True`

.. code:: ada project=Courses.Ada_For_C_Embedded_Dev.SPARK.Exceptions

    procedure P is
    begin
       Statements_That_Might_Raise_Exceptions;
    exception
       when Exception_A =>
          Handle_Exception_A;
       when Exception_B =>
          Handle_Exception_B;
       when Exception_C =>
          Handle_Exception_C;
    end P;

The name :ada:`Exception_A` refers to a specific visible exception
declared somewhere with that name, as do the other two exceptions
referenced. Each handler consists of an arbitrary sequence of
statements, in this case specific procedures called in response to those

> specific exceptions.

Reviewer: Another suggestion:
```suggestion
specific exceptions. For example, when handling :ada:`Exception_A`, we call :ada:`Handle_Exception_A`.
```
I know, this is pretty obvious when reading the code. Sometimes, however, it's better to add those details just to ensure all readers are understanding the context.

Note how the syntactic structure of the exception handling section
resembles that of a case statement. The resemblance is intentional, to
suggest similar behavior. When something in the statements of the normal
execution raises an exception, normal execution is abandoned and the
corresponding exception handler for that specific exception is executed.
Only that one handler is executed and then the routine completes. The

> handlers do not "fall through" to the handlers below, again like the

Reviewer: Yet another suggestions
```suggestion
handlers do not "fall through" to the handlers below. In the code above, as soon as :ada:`Exception_A` has been handled, there's no handling of :ada:`Exception_B`. This is like the
```

> behavior of case statements.

Reviewer: ```suggestion
behavior of case statements in Ada |mdash| there's no need to use a
:c:`break` statement as in C.
```

The semantic model is the same as with many other programming languages,
in which normal execution is unavoidably interrupted in the face of an
active exception. This design precludes missing (intentionally or
otherwise) a function status result indicating an error, thus giving the
developer a strong guarantee that errors are detected and addressed.

> When no corresponding handler is located in the exception handlers 

Reviewer: Editorial suggestion: I'm not sure I'd understand this paragraph if I didn't know how exceptions are supposed to work. Maybe the whole paragraph could be merged with the paragraph below that explains the source-code example.

    with Arrays;      use Arrays;

    procedure Some_Process is
       L : constant List (1 .. 100) := (others => 42);
    begin
       Put_Line (Integer'Image (Value (L, 1, 10)));
    exception
       when Constraint_Error =>
          Put_Line ("Constraint_Error caught in Some_Process");
          Put_Line ("Some_Process completes normally");
    end Some_Process;

    with Some_Process;
    with Ada.Text_IO; use Ada.Text_IO;

    procedure Main is
    begin
       Some_Process;
       Put_Line ("Main completes normally");
    end Main;

:code-config:`accumulate_code=False`

:code-config:`reset_accumulator=True`

> Procedure :ada:`Main` calls :ada:`Some_Process`, which in turn calls 

Reviewer: Editorial suggestion: this paragraph is actually a perfect description of how the exception is propagated and handled in the code above. And yet, I had to read it carefully to understand it. Maybe we could split the paragraph in bullet points for each subprogram in the chain? Another (more complicated) approach would be to add a graph that illustrates the chain and shows where the exception is raised, propagated and handled.

function :ada:`Value` (line 7). The call to :ada:`Value` has arguments 
leading to an attempt to access an array component via an out-of-bounds 
index (:ada:`1 + 10 * 10 = 101`, beyond the last index of :ada:`L`). 
This attempt will trigger an exception in :ada:`Value` prior to actually 
accessing the array object's memory. Function :ada:`Value` doesn't have 
any exception handlers so the exception is propagated up to the caller 
:ada:`Some_Process`. Procedure :ada:`Some_Process` has an exception 
handler for :ada:`Constraint_Error` and it so happens that that
:ada:`Constraint_Error` is the exception raised in this case. As a 
result, the code for that handler will be executed, printing some 
messages on the screen. Then procedure :ada:`Some_Process` will return 
to :ada:`Main` normally. :ada:`Main` then continues to execute normally 
after the call to :ada:`Some_Process` and prints its completion message. 

If procedure :ada:`Some_Process` had also not had a matching exception 

> handler for :ada:`Constraint_Error` that procedure call would also have 

Reviewer: Editorial remark: I believe there should be a comma after `Constraint_Error`.

completed abnormally, immediately, without printing its closing message. 
You can try that situation by removing the exception section in 
:ada:`Some_Process` and re-running the code. 

Full exception propagation as described above is not the norm for 
embedded applications when the highest levels of integrity are required. 
The run-time library code implementing exception propagation can be 
rather complex and expensive to certify, as is the application code when 
control flow can traverse dynamically up the call chain. Instead of the 
full exception model, high-integrity applications often take alternative 
approaches. 

One alternative consists of deactivating exceptions altogether, or more 
precisely, deactivating language-defined checks, which means that the 
compiler will not generate code checking for conditions giving rise to 
exceptions. Of course, this makes the code vulnerable to attacks, such 
as buffer overflow, unless otherwise verified (e.g. through static 
analysis). Deactivation can be applied at the unit level, through the 
``-gnatp`` compiler switch, or locally within a unit via the 
:ada:`pragma Suppress`. For example, we can write the following. Note 

> the pragma on line 4 of "arrays.adb": 

Reviewer: _Editorial remark_: you can use this syntax here:
```
:file:`arrays.adb`
```

    void __gnat_last_chance_handler (char *source_location,
                                     int line);

We'll go into the details of the :ada:`pragma Export` in a further 
section on language interfacing. For now, just know that the symbol 

> "__gnat_last_chance_handler" is what the run-time uses to branch 

Reviewer: Editorial remark: you can use this syntax:
```
:c:`__gnat_last_chance_handler`
```

> immediately to the last-chance handler. Pragma Export associates that 

Reviewer: ... and you can use this syntax here:
```
:ada:`Export`
```

symbol with this replacement procedure so it will be invoked instead of 
the default routine. As a result, the actual procedure name in Ada is 
immaterial. 

Here is an example implementation that simply blinks an LED
forever on the target:

.. code-block:: ada

>    procedure Last_Chance_Handler (Msg : System.Address; Line : Integer) is

Reviewer: Very nice example, thanks for it!

      pragma Unreferenced (Msg, Line);

      Next_Release : Time := Clock;
      Period       : constant Time_Span := Milliseconds (500);
   begin
      Initialize_LEDs;
      All_LEDs_Off;

      loop
         Toggle (LCH_LED);
         Next_Release := Next_Release + Period;
         delay until Next_Release;
      end loop;
   end Last_Chance_Handler;

The :ada:`LCH_LED` is a constant referencing the LED used by the
last-chance handler, declared elsewhere. The infinite loop is necessary
because a last-chance handler must never return to the caller (hence the
term "last-chance"). The LED changes state every half-second.

Unlike the approach in which there is only the last-chance handler 
routine, the other approach also allows exception handlers but in a 

> specific idiomatic manner. In particular, whenever an exception is 

Reviewer: Editorial suggestion: maybe we should also point to documentation that describes how to configure the compiler to use these two approaches.

Reviewer: (GitHub won't let me reply to your question below...)

To create links, you have to use this format:
```
`LINK_TITLE <WWW_LINK>`_
```

Just as an example, the text below includes a link to the GPRbuild documentation:
```
Refer to the `GPRbuild User’s Guide <https://docs.adacore.com/gprbuild-docs/html/gprbuild_ug.html>`_ for more details.
```

the Ada reserved word :ada:`others`. As in case statements, it covers
all other choices not explicitly mentioned, and so much come last. For
example:

.. code:: ada run_button project=Courses.Ada_For_C_Embedded_Dev.SPARK.Exception_Return_Others

    package Arrays is

       type List is array (Natural range <>) of Integer;

       function Value (A : List; X, Y : Integer) return Integer;

    end Arrays;

    package body Arrays is

       function Value (A : List; X, Y : Integer) return Integer is
       begin
          return A (X + Y * 10);
       exception
          when Constraint_Error =>
             return 0;
          when others =>
             return -1;
       end Value;

    end Arrays;

    with Ada.Text_IO; use Ada.Text_IO;
    with Arrays;      use Arrays;

    procedure Some_Process is
       L : constant List (1 .. 100) := (others => 42);
    begin
       Put_Line (Integer'Image (Value (L, 1, 10)));
    exception
       when Constraint_Error =>
          Put_Line ("FAILURE");
    end;

In the code above, the :ada:`Value` function has a handler specifically 
for :ada:`Constraint_Error` as before, but also now has a handler for 

> all other exceptions. For any exception other than Constraint_Error 

Reviewer: Editorial comment: I'd use the `:ada:` syntax and add a comma:
```
:ada:`Constraint_Error`,
```

handler will catch the exception and -1 will be returned instead of zero. 

Understanding Dynamic Checks versus Formal Proof
------------------------------------------------

So far, we have discussed language-defined checks inserted by the
compiler for verification at run-time. We saw that these dynamic checks
verified semantic conditions ensuring proper execution, such as
preventing writing past the end of a buffer, or exceeding an
application-specific integer range constraint, and so on. These checks
are defined by the language because they apply generally and can be
expressed in language-defined terms.

*Developers* can also define dynamic checks. These checks specify 
component-specific or application-specific conditions, expressed in 
terms defined by the component or application. We will refer to these 
checks as "user-defined" for convenience. 

Like the language-defined checks, user-defined checks must be 
true at run-time. All checks consist of boolean conditions, which is why 
we can refer to them as assertions: their conditions are asserted to be 
true by the developer and can be verified at run-time. 

Assertions come in several forms, some relatively low-level,

> such as a simple pragma Assert, and some high-level, such as type

Reviewer: Editorial remark: I'd write:
```
pragma :ada:`Assert`
```

       Put_Line (Integer'Image (X));
    end Main;

If we compiled this code we would get a warning about the assignment after the 
pragma Assert, but not one about the Assert itself on line 7. 

::

   gprbuild -q -P main.gpr
   main.adb:8:11: warning: value not in range of type "Standard.Positive"
   main.adb:8:11: warning: "Constraint_Error" will be raised at run time

No code is generated for the user-defined check expressed via pragma 

> Assert but the language-defined check is emitted. In this case the range 

Reviewer: Editorial remark:
```
pragma :ada:`Assert`
```

       Put_Line (Integer'Image (X));
    end Main;

Now we also get the compiler warning about the pragma Assert condition. 
When run, the failure of pragma Assert on line 7 raises the exception 
Ada.Assertions.Assertion_Error. According to the expression in the 
assertion, :ada:`X` is expected (incorrectly) to be above 99 after the 
multiplication. (The exception name in the error message, 
SYSTEM.ASSERTIONS.ASSERT_FAILURE, is a GNAT-specific alias for 
Ada.Assertions.Assertion_Error.) 

It's interesting to see in the output that the compiler can detect some 
violations at compile-time: 

::

    main.adb:7:19: warning: assertion will fail at run time
    main.adb:7:21: warning: condition can only be True if invalid values present
    main.adb:8:11: warning: value not in range of type "Standard.Positive"

Generally speaking, a complete analysis of that sort is beyond the scope 
of compilers and they will not find all errors prior to execution, even 

> ones we can detect ourselves by inspection. More errors can be found by 

Reviewer: Question by a non-native speaker: shouldn't it be "even _the_ ones we can detect..."?

either through implicit language-defined checks (is the index in the 
right range?) or explicit user-defined checks. These checks are already 
useful by themselves but they have an important limitation: the assertions 
are in the implementation, hidden from the callers of the routine. For 
example, a call's success or failure may depend upon certain input 
values but the caller doesn't have that information. 

Generally speaking, Ada and SPARK put a lot of emphasis on strong,
complete specifications for the sake of abstraction and analysis.
Callers need not examine the implementations to determine
whether the arguments passed to it are changed, for example. It is
possible to go beyond that, however, to specify arbitrary implementation
constraints and behavior requirements via contracts.

At the language level, contracts are higher-level forms of assertions
associated with specifications and declarations, rather than sequences
of statements. Like other assertions they can be activated or
deactivated at run-time, and can be statically proven. We'll concentrate
here on two kinds of contracts, both associated especially (but not
exclusively) with procedures and functions:

> - *Preconditions*, those boolean conditions expected to be true *prior* to the call of a given subprogram 

Reviewer: Editorial remark: I usually write "Boolean" instead of "boolean" because it refers to George Boole.

the user-defined assertions correctly and sufficiently describe and 
constrain the implementation of the corresponding routines. 

Although we are not proving whole program correctness, as you will have 
seen |mdash| and done |mdash| we can prove properties than make our 
software far more robust and bug-free than is possible otherwise. But in 
addition, consider what proving the unit-level requirements for your 
procedures and functions would do for the cost of unit testing and 
system integration. The tests would pass the first time. 

However, within the scope of what SPARK can do, not everything can be 
proven. In some cases that is because the software behavior is not 
amenable to expression as boolean conditions (for example, a mouse 
driver). In other cases the source code is beyond the capabilities of 
the analyzers that actually do the mathematical proof. In these cases 
the combination of proof and actual test is appropriate, and still less 
expensive that testing alone. 

There is, of course, much more to be said about what can be done with 
SPARK and :program:`gnatprove`. Those topics are reserved for the 

> advanced SPARK course. 

Reviewer: I think we should link to the existing _Intro to SPARK_ course instead. This would be the appropriate syntax:
```
:doc:`Introduction to SPARK <courses/intro-to-spark/index>` course.
```
